### ISSUE TITLE
FileSaver replaces LF with CRLF

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1979]
|<img alt="agreif's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| agreif reported|
|-|-|
|Trac ID|trac#1979|
|Type|defect|
|Created|2006-08-01 07:14:50Z|

On windows XP (python 2.4.2) the latest svn checkout produces the following bug:
FileSaver converts LF to CRLF automatically, even in binary files!!

I tried to upload files with the following code:

 child_uploadfile = static.FileSaver(destination='d:/agreif/tmp',
                                               expectedFields=['filename'],
                                               maxBytes=10000000,

allowedTypes=[http_headers.MimeType('text', 'x-python'),
              http_headers.MimeType('application','x-zip-compressed')])

here is the hexdump of two samples.
You see that in front of every occurrance of "0A" a "0D" is inserted.

Sample 1: textfile
orig:
68616C6C6F0A
uploaded:
68616C6C6F0D0A

Sample 2 binary file (zip)
orig:
504B03040A00000000001589FF34D13172B90500000005000000100000004E6575205465787464617465692E707968616C6C6F504B010214000A00000000001589FF34D13172B905000000050000001000000000000000010020000000000000004E6575205465787464617465692E7079504B050600000000010001003E000000330000000000
uploaded:
504B03040D0A00000000001589FF34D13172B90500000005000000100000004E6575205465787464617465692E707968616C6C6F504B010214000D0A00000000001589FF34D13172B905000000050000001000000000000000010020000000000000004E6575205465787464617465692E7079504B050600000000010001003E000000330000000000

Attachments:

* [1979_test.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e41/e411a490148911bb6ee16beaa6f794a437b3bc9d/687c8026b8c2e7c4ab1b8241181f098986a8a7fe.txt) (829 bytes) - added by moonfallen on 2006-08-01 15:40:54Z - Failing test in test_static to expose this problem

<details><summary>Searchable metadata</summary>

```
trac-id__1979 1979
type__defect defect
reporter__agreif agreif
priority__high high
milestone__Web2_Gold_Master Web2-Gold-Master
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1154416490000000 1154416490000000
changetime__1181939589000000 1181939589000000
version__None None
owner__dialtone dialtone
cc__alex.greif@... cc__moonfallen cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

I thought this would be a simple matter of changing 'w' to 'wb' in static.py, but when I made that change the test still fails, so the stream is producing bad data.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Since file uploads are such a major new feature of web2 (ostensibly the reason for its existence) I'd say that this is important to its taking over from web...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

This problem seems to exist under windows only.

**Comment by twisted-trac** [comment]
|<img alt="dialtone's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|dialtone commented|
|-|-|

I think I fixed this bug in filesaver-1979 branch. It's now ready for review.

**Comment by twisted-trac** [comment]
|<img alt="dialtone's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|dialtone commented|
|-|-|

small changes but still set for review

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_8"></a>|@jyknight set owner to dialtone|
|-|-|

Looks fine on windows now; I noticed that docstring for writeFile mentions the nonexistent isWriteable instead of isSafeToWrite.

**Comment by twisted-trac** [comment]
|<img alt="dialtone's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|dialtone set status to `closed`|
|-|-|

(In [20546]) static.FileSaver() doesn't add '\r\n' when run on windows anymore.

Author: dialtone

Reviewer: foom

Fixes [#1979](https://github.com/twisted/twisted/issues/1979)
